src/backend/distributed/commands/dependencies.c

```diff
@@ -0,0 +1,196 @@
+/*-------------------------------------------------------------------------
+ *
+ * dependencies.c
+ *    Commands to create dependencies of an object on all workers.
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "catalog/dependency.h"
+#include "catalog/objectaddress.h"
+
+#include "distributed/dist_catalog/dependency.h"
+#include "distributed/connection_management.h"
+#include "distributed/worker_manager.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/commands.h"
+#include "distributed/remote_commands.h"
+#include "distributed/dist_catalog/distobject.h"
+
+static List * GetDependencyCreateDDLCommands(const ObjectAddress *dependency);
+
+
+/*
+ * EnsureDependenciesExists finds all the dependencies that we support and makes sure
+ * these are available on all workers. If not available they will be created on the
+ * workers via a separate session that will be committed directly so that the objects are
+ * visible to potentially multiple sessions creating the shards.
+ *
+ * Note; only the actual objects are created via a separate session, the local records to
+ * pg_dist_object are created in this session. As a side effect the objects could be
+ * created on the workers without a catalog entry on the coordinator. Updates to the
+ * objects on the coordinator are not propagated to the workers until the record is
+ * visible on the coordinator.
+ *
+ * This is solved by creating the dependencies in an idempotent manner, either via
+ * postgres native CREATE IF NOT EXISTS, or citus helper functions.
+ */
+void
+EnsureDependenciesExistsOnAllNodes(const ObjectAddress *target)
+{
+	const uint64 connectionFlag = FORCE_NEW_CONNECTION;
+
+	/* local variables to work with dependencies */
+	List *dependencies = NIL;
+	ListCell *dependencyCell = NULL;
+
+	/* local variables to collect ddl commands */
+	List *ddlCommands = NULL;
+
+	/* local variables to work with worker nodes */
+	List *workerNodeList = NULL;
+	ListCell *workerNodeCell = NULL;
+	List *connections = NULL;
+	ListCell *connectionCell = NULL;
+
+	/*
+	 * collect all dependencies in creation order and get their ddl commands
+	 */
+	dependencies = GetDependenciesForObject(target);
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		ddlCommands = list_concat(ddlCommands,
+								  GetDependencyCreateDDLCommands(dependency));
+	}
+	if (list_length(ddlCommands) <= 0)
+	{
+		/* no ddl commands to be executed */
+		return;
+	}
+
+	/*
+	 * collect and connect to all applicable nodes
+	 */
+	workerNodeList = ActivePrimaryNodeList();
```

Comment:
If there is an ongoing master_add/update/remove_node, then the new node will be added but won't have the objects. That is basically equivalent to the master_add_node happening before this point, but if we expect master_add_node to also create the objects then we need to make sure we include the new node here. 

To achieve that, we need to take a lock that conflicts with master_add_node. We also need to make sure we keep that lock until the end of the current transaction, since that's when the pg_dist_object records gets committed. If a master_add_node starts after we take the lock, it will block until the current transaction is released and then the master_add_node will see the new type in pg_dist_object.

Both master_add_node and master_remove_node take an exclusive lock on pg_dist_node. master_update_node only takes a row exclusive lock, but we need not care about it, since we only expect it to be used to change to an exact replica. 

I suggest taking a row share lock, which [conflicts](https://www.postgresql.org/docs/current/explicit-locking.html) with the exclusive locks:
```c
LockRelationOid(DistNodeRelationId(), RowShareLock);
```

Note that the node metadata cache will be invalidated as soon as we obtain the lock, since the lock acquisition functions call `AcceptInvalidationMessages()`, which reads invalidation messages that were sent when the master_add_node committed from a shared memory queue and calls the appropriate invalidation handlers in Citus code.

We also need to make sure we do not read from pg_dist_object in master_add_node until after [we take the exclusive lock].(https://github.com/citusdata/citus/blob/master/src/backend/distributed/utils/node_metadata.c#L450).

I looked at how the existing locking for creating a reference table concurrently with master_add_node works, but apparently there is none so you end up with an active node without a reference table copy. That'll need to be fixed separately.

Reply:
This lock worker pretty well here, also added isolation tests to verify the behaviour